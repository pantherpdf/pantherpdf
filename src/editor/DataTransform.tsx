/**
 * DataTransform.tsx
 * Manage data transformations. Add, delete, edit Transform.
 */

import React, { useState } from 'react';
import { Report, TransformItem } from '../types';
import { GeneralProps, Transform } from './types';
import { Dropdown, Button, ButtonGroup, Modal } from 'react-bootstrap';
import Trans, { TransName } from '../translation';
import { getTransform } from '../transforms/allTransforms';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faArrowDown,
  faArrowUp,
  faEdit,
  faFolderOpen,
  faPlus,
  faTimes,
  faTrash,
} from '@fortawesome/free-solid-svg-icons';
import ObjectExplorer from './ObjectExplorer';
import InputApplyOnEnter from '../widgets/InputApplyOnEnter';
import globalStyle from '../globalStyle.module.css';

/**
 * Function for calling all transformations
 * @func
 * @param {inputData} unknown - Input data
 * @param {len} number - Number of transformations to apply
 */
export async function transformData(
  allTrans: Transform[],
  inputData: unknown,
  transformData: TransformItem[],
  len?: number,
) {
  if (len === undefined) {
    len = transformData.length;
  }
  for (let i = 0; i < len; ++i) {
    const w = transformData[i];
    const comp = getTransform(allTrans, w.type);
    inputData = await comp.transform(inputData, w);
  }
  return inputData;
}

interface TransformItemProps extends GeneralProps {
  item: TransformItem;
  index: number;
  showData: (len: number) => void;
  up: (idx: number) => void;
  down: (idx: number) => void;
  itemDelete: (idx: number) => void;
  openEditor: (idx: number) => void;
}
function TransformItemEditor(props: TransformItemProps) {
  const comp = getTransform(props.transforms, props.item.type);
  return (
    <Dropdown as={ButtonGroup} className="d-flex" size="sm">
      <button
        className="btn btn-sm btn-outline-secondary flex-fill text-left"
        onClick={() => props.showData(props.index + 1)}
      >
        <span className="d-block">{TransName(comp.name)}</span>
        <small
          className="d-block text-truncate text-nowrap text-muted"
          style={{ maxWidth: '10rem' }}
        >
          {props.item.comment}
        </small>
      </button>
      <Dropdown.Toggle
        style={{ maxWidth: '3rem' }}
        split
        variant="outline-secondary"
        id="dropdown-custom-2"
      />
      <Dropdown.Menu>
        <Dropdown.Item onClick={() => props.openEditor(props.index)}>
          <FontAwesomeIcon icon={faEdit} className="me-2" fixedWidth />
          {Trans('edit')}
        </Dropdown.Item>
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.up(props.index);
          }}
          disabled={props.index === 0}
        >
          <FontAwesomeIcon icon={faArrowUp} className="me-2" fixedWidth />
          {Trans('up')}
        </Dropdown.Item>
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.down(props.index);
          }}
          disabled={props.index + 1 === props.report.transforms.length}
        >
          <FontAwesomeIcon icon={faArrowDown} className="me-2" fixedWidth />
          {Trans('down')}
        </Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item
          onClick={e => {
            e.preventDefault();
            props.itemDelete(props.index);
          }}
        >
          <FontAwesomeIcon icon={faTrash} className="me-2" fixedWidth />
          {Trans('delete')}
        </Dropdown.Item>
      </Dropdown.Menu>
    </Dropdown>
  );
}

async function uploadOverrideSourceData(): Promise<unknown> {
  return new Promise(resolve => {
    let el = document.createElement('INPUT') as HTMLInputElement;
    el.type = 'file';
    el.accept = 'application/json';
    el.multiple = false;
    el.addEventListener('change', function (ev2) {
      const f = el.files && el.files.length > 0 ? el.files[0] : undefined;
      if (!f) {
        return resolve(undefined);
      }
      try {
        const reader = new FileReader();
        reader.onload = e2 => {
          if (!e2.target || typeof e2.target.result !== 'string') {
            return;
          }
          const dt = JSON.parse(e2.target.result);
          resolve(dt);
        };
        reader.readAsText(f);
      } catch (e) {
        alert(`Error: ${String(e)}`);
        return resolve(undefined);
      }
    });
    el.click();
  });
}

interface TEdit {
  index: number;
  data: TransformItem;
}

interface SourceDataShow {
  length?: number;
  data: unknown;
  errorMsg?: string;
}

export default function DataTransform(props: GeneralProps) {
  const [shownModalInsert, setShownModalInsert] = useState<boolean>(false);
  const [showEdit, setShowEdit] = useState<TEdit | null>(null);
  const [showData, setShowData] = useState<SourceDataShow | null>(null);

  async function doShowData(len: number) {
    let dt;
    try {
      dt = await props.getSourceData();
    } catch (e) {
      let msg = String(e);
      if (msg.trim().length === 0) {
        msg = 'unknown error';
      }
      setShowData({ data: null, errorMsg: msg });
      return;
    }

    let dt2;
    try {
      dt2 = await transformData(
        props.transforms,
        dt,
        props.report.transforms,
        len,
      );
    } catch (e) {
      let msg = String(e);
      if (msg.trim().length === 0) {
        msg = 'unknown error';
      }
      setShowData({ data: null, errorMsg: msg });
      return;
    }
    setShowData({ data: dt2, length: len });
  }

  async function itemAdd(key: string) {
    const cmp = getTransform(props.transforms, key);
    const dt = await cmp.newItem();
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms, dt],
    };
    await props.setReport(report2);
    setShownModalInsert(false);
    setShowEdit({
      index: report2.transforms.length - 1,
      data: JSON.parse(JSON.stringify(dt)),
    });
  }

  async function itemUp(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    const tmp = report2.transforms[idx - 1];
    report2.transforms[idx - 1] = report2.transforms[idx];
    report2.transforms[idx] = tmp;
    return props.setReport(report2);
  }

  async function itemDown(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    const tmp = report2.transforms[idx + 1];
    report2.transforms[idx + 1] = report2.transforms[idx];
    report2.transforms[idx] = tmp;
    return props.setReport(report2);
  }

  async function itemDelete(idx: number) {
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    report2.transforms.splice(idx, 1);
    return props.setReport(report2);
  }

  function itemOpenEditor(idx: number) {
    setShowEdit({
      index: idx,
      data: JSON.parse(JSON.stringify(props.report.transforms[idx])),
    });
  }

  async function editSave() {
    if (!showEdit) {
      return;
    }
    const report2: Report = {
      ...props.report,
      transforms: [...props.report.transforms],
    };
    report2.transforms[showEdit.index] = showEdit.data;
    await props.setReport(report2);
    setShowEdit(null);
  }

  function editChange(data: TransformItem) {
    if (showEdit) {
      setShowEdit({ ...showEdit, data });
    }
  }

  const editCmp = showEdit
    ? getTransform(props.transforms, showEdit.data.type)
    : null;

  return (
    <>
      <div className={`${globalStyle.section} d-flex`}>
        <label htmlFor="source-url" className="flex-fill">
          {Trans('source data')}
        </label>
        {!!props.setSourceDataOverride && (
          <button
            className="btn btn-link py-0"
            style={{ color: '#666' }}
            onClick={async () => {
              if (!props.setSourceDataOverride) {
                return;
              }
              if (props.isSourceDataOverriden) {
                // remove
                props.setSourceDataOverride(undefined);
              } else {
                // select file and replace
                const dt = await uploadOverrideSourceData();
                if (dt) {
                  props.setSourceDataOverride(dt);
                }
              }
            }}
            title={Trans('load local json file')}
          >
            <FontAwesomeIcon
              icon={props.isSourceDataOverriden ? faTimes : faFolderOpen}
            />
          </button>
        )}
      </div>
      <InputApplyOnEnter
        id="source-url"
        value={props.report.dataUrl}
        onChange={val => {
          const report2: Report = { ...props.report, dataUrl: String(val) };
          props.setReport(report2);
        }}
        placeholder="https://www.example.com/data.js(on)"
      />

      <div className={globalStyle.section}>{Trans('transform')}</div>
      {props.report.transforms.map((item, idx) => (
        <TransformItemEditor
          {...props}
          item={item}
          index={idx}
          key={idx}
          showData={doShowData}
          up={itemUp}
          down={itemDown}
          itemDelete={itemDelete}
          openEditor={itemOpenEditor}
        />
      ))}
      <div className="">
        <button
          className="btn btn-sm btn-outline-secondary mt-2"
          onClick={() => setShownModalInsert(true)}
        >
          <FontAwesomeIcon icon={faPlus} className="me-2" />
          {Trans('transform')}
        </button>
      </div>

      {/* ADD NEW */}
      <Modal show={shownModalInsert} onHide={() => setShownModalInsert(false)}>
        <Modal.Header closeButton>
          <Modal.Title>{Trans('insert data transform')}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div className="list-group">
            {props.transforms.map(w => {
              return (
                <button
                  className="btn list-group-item list-group-item-action"
                  key={w.id}
                  onClick={() => itemAdd(w.id)}
                >
                  {TransName(w.name)}
                </button>
              );
            })}
          </div>
        </Modal.Body>
      </Modal>

      {/* EDIT */}
      <Modal show={!!showEdit} onHide={() => setShowEdit(null)}>
        {editCmp && showEdit && (
          <>
            <Modal.Header closeButton>
              <Modal.Title>
                {Trans('edit')} {TransName(editCmp.name)}
              </Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <editCmp.RenderEditor
                {...props}
                index={showEdit.index}
                item={showEdit.data}
                setItem={editChange}
              />
            </Modal.Body>
            <Modal.Footer>
              <input
                type="text"
                className="form-control flex-fill w-auto"
                id="trans-edit-comment"
                value={showEdit.data.comment}
                placeholder={Trans('comment')}
                onChange={e =>
                  editChange({
                    ...showEdit.data,
                    comment: e.currentTarget.value,
                  })
                }
              />
              <Button variant="secondary" onClick={() => setShowEdit(null)}>
                {Trans('cancel')}
              </Button>
              <Button variant="secondary" onClick={editSave}>
                {Trans('save')}
              </Button>
            </Modal.Footer>
          </>
        )}
      </Modal>

      {/* DATA */}
      <Modal show={!!showData} onHide={() => setShowData(null)}>
        <Modal.Header closeButton>
          <Modal.Title>{Trans('data')}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <>
            {showData?.data && <ObjectExplorer data={showData.data} />}
            {showData?.errorMsg && <div>{showData?.errorMsg}</div>}
          </>
        </Modal.Body>
      </Modal>
    </>
  );
}
