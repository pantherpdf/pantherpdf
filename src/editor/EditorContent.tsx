/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import React, { useEffect } from 'react';
import Typography from '@mui/material/Typography';
import { defaultReportCss } from '../types';
import { propertyFontGenCss } from '../widgets/PropertyFont';
import type { GeneralProps } from './types';
import trans from '../translation';
import getFontsUsed from '../data/getFontsUsed';

export default function EditorContent(props: GeneralProps) {
  useEffect(() => {
    const fontsUsed = getFontsUsed(props.report, props.widgets);
    const urls = props.api.fonts?.getCssUrls(fontsUsed) || [];
    loadFonts(urls);
  });
  const style = {
    ...defaultReportCss,
    ...propertyFontGenCss(props.report.properties.font || {}),
  };
  const width = props.report.properties.paper?.width || 210;
  style.maxWidth = `${(width * 800) / 210}px`;
  style.margin = `0 auto`;
  return (
    <div style={style}>
      {props.renderWidgets(props.report.widgets, [])}
      {props.report.widgets.length === 0 && (
        <Typography color="GrayText" align="center">
          {trans('drag drop widgets here')}
        </Typography>
      )}
    </div>
  );
}

const loadedUrls: string[] = [];
function loadFonts(urls: string[]): void {
  for (const url of urls) {
    if (loadedUrls.indexOf(url) !== -1) {
      continue;
    }
    loadedUrls.push(url);
    const link = window.document.createElement('link');
    link.rel = 'stylesheet';
    link.href = url;
    window.document.head.appendChild(link);
  }
}
