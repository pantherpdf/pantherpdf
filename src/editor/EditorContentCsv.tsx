/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableRow from '@mui/material/TableRow';
import TableCell from '@mui/material/TableCell';
import Trans from '../translation';
import ErrorAlert from './ErrorAlert';
import type { GeneralProps } from './types';
import useTransformedData from './useTransformedData';

export default function RenderContentCsv(props: GeneralProps) {
  const data = useTransformedData(props);
  if (!data.ok) {
    return <ErrorAlert msg={data.errorMsg} />;
  }
  if (!Array.isArray(data.value)) {
    return <ErrorAlert msg={Trans('data must be 2D array')} />;
  }
  return (
    <Table>
      <TableBody>
        {data.value.map((row, idx) => (
          <TableRow key={idx}>
            {Array.isArray(row) &&
              row.map((cell, idx2) => (
                <TableCell key={idx2}>{String(cell)}</TableCell>
              ))}
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
}
