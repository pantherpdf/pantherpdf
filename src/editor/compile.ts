/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021
 * @license MIT
 */

import { Report, Item, ApiEndpoints } from '../types';
import type { ReportCompiled } from '../types';
import FormulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import { CompileHelper, Widget } from './types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: any } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  const getVar = formulaHelper.getVar.bind(formulaHelper);
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = await FormulaEvaluate(v.formula, { getVar });
    formulaHelper.push(v.name, getVarValue);
  }

  let dt2: ReportCompiled = {
    ...report,
    children: [],
    fontsUsed: [],
    globalCss: '',
  };
  dt2 = JSON.parse(JSON.stringify(dt2));
  if (dt2.properties.font) {
    const style = PropertyFontExtractStyle(dt2.properties.font);
    if (style) {
      dt2.fontsUsed.push(style);
    }
  }

  if (dt2.properties.fileName) {
    const res = await FormulaEvaluate(dt2.properties.fileName, { getVar });
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    dt2.properties.fileName = res;
  }

  const helper: CompileHelper = {
    wid: [],
    report: report,
    reportCompiled: dt2,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: async (txt: string) => {
      return FormulaEvaluate(txt, { getVar });
    },

    compileChildren: async (children: Item[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  dt2.children = await helper.compileChildren(report.children, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }
  return dt2;
}
