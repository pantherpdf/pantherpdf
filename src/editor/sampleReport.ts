/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import type { Report } from '../types';

export const sampleReport: Report = {
  name: 'Report 1',
  widgets: [],
  transforms: [],
  properties: {},
  variables: [],
};
