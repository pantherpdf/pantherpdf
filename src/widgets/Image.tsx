/**
 * Image.tsx
 */

import React, { useState, CSSProperties } from 'react';
import { TData, TDataCompiled } from '../types';
import type { Widget } from '../editor/types';
import { faEllipsisH, faImage } from '@fortawesome/free-solid-svg-icons';
import BoxName from './BoxName';
import InputApplyOnEnter, {
  WidthOptions,
  WidthRegex,
} from './InputApplyOnEnter';
import PropertyAlign, { TAlign } from './PropertyAlign';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Modal } from 'react-bootstrap';
import FileDialog from '../FileDialog';
import Trans from '../translation';
import base64ArrayBuffer from './base64ArrayBuffer';
import globalStyle from '../globalStyle.module.css';

export interface ImageData extends TData {
  type: 'Image';
  url: string;
  formula: string;
  align?: TAlign;
  width?: string;
  height?: string;
  fit?: 'fill' | 'contain' | 'cover';
}

export interface ImageCompiled extends TDataCompiled {
  type: 'Image';
  data: string;
  align?: TAlign;
  width?: string;
  height?: string;
  fit?: 'fill' | 'contain' | 'cover';
}

// to force reload images when user reuploads or deletes image
// this will append &state=<number> to url to force reload
let imgState = 0;

export const Image: Widget = {
  id: 'Image',
  name: { en: 'Image', sl: 'Slika' },
  icon: { fontawesome: faImage },

  newItem: async (): Promise<ImageData> => {
    return {
      type: 'Image',
      children: [],
      url: '',
      formula: '',
      width: '',
    };
  },

  compile: async (dt: ImageData, helpers): Promise<ImageCompiled> => {
    let data;
    if (dt.url.length > 0) {
      data = dt.url;
      if (data.startsWith('local/')) {
        if (!helpers.api.filesDownload) {
          throw new Error('Missing api.filesDownload');
        }
        const obj = await helpers.api.filesDownload(data.substring(6));
        if (obj.mimeType.indexOf(';') !== -1) {
          throw new Error('Bad mime type');
        }
        data = `data:${obj.mimeType};base64,${base64ArrayBuffer(obj.data)}`;
      }
    }
    //
    else if (dt.formula.length > 0) {
      const data2 = await helpers.evalFormula(dt.formula);
      if (typeof data2 !== 'string') {
        throw new Error(
          `Image from formula expected string but got ${typeof data2}`,
        );
      }
      data = data2;
    }
    //
    else {
      data = '';
    }
    return {
      type: dt.type,
      data,
      align: dt.align,
      width: dt.width,
      height: dt.height,
      fit: dt.fit,
    };
  },

  Render: function (props) {
    const item = props.item as ImageData;

    const cssImg: CSSProperties = {
      display: 'inline-block',
    };
    if (item.width) {
      cssImg.width = item.width;
    } else {
      cssImg.maxWidth = '100%';
    }

    if (item.height) {
      cssImg.height = item.height;
    }
    if (item.width && item.height) {
      cssImg.objectFit = item.fit || 'fill';
    }

    let img;
    if (item.url.length > 0) {
      let url = item.url;
      if (url.startsWith('local/')) {
        url = props.api.filesDownloadUrl
          ? props.api.filesDownloadUrl(url.substring(6))
          : '';
        // force refresh
        const hasQuestion = url.indexOf('?') !== -1;
        url += (hasQuestion ? '&' : '?') + '_reports_state=' + String(imgState);
      }
      img = <img src={url} alt="" style={cssImg} />;
    }
    //
    else if (item.formula.length > 0) {
      cssImg.backgroundColor = '#ddd';
      cssImg.textAlign = 'center';
      cssImg.minWidth = '100px';
      cssImg.padding = '20px 0';
      cssImg.overflow = 'hidden';
      img = (
        <div style={cssImg}>
          <span
            className="font-monospace text-nowrap overflow-hidden"
            style={{ fontSize: '70%', opacity: '0.3' }}
          >
            {item.formula}
          </span>
        </div>
      );
    }
    //
    else {
      img = (
        <small className="d-block text-muted">
          {Trans('no image selected')}
        </small>
      );
    }

    const cssContainer: CSSProperties = {};
    if (item.align) {
      cssContainer.textAlign = item.align;
    }

    return (
      <BoxName {...props} name={Image.name}>
        <div style={cssContainer}>{img}</div>
      </BoxName>
    );
  },

  RenderFinal: function (props) {
    const item = props.item as ImageCompiled;
    const cssImg: CSSProperties = {
      display: 'inline-block',
    };
    if (item.width) {
      cssImg.width = item.width;
    } else {
      cssImg.maxWidth = '100%';
    }
    if (item.height) {
      cssImg.height = item.height;
    }
    if (item.width && item.height) {
      cssImg.objectFit = item.fit || 'fill';
    }
    const cssContainer: CSSProperties = {};
    if (item.align) {
      cssContainer.textAlign = item.align;
    }

    let img;
    if (item.data === '') {
      return '';
    }
    //
    else if (item.data.trimStart().startsWith('<svg')) {
      img = `<div style="${props.styleToStringAttribute(cssImg)}">${
        item.data
      }</div>`;
    }
    //
    else if (
      item.data.startsWith('data:image/') ||
      item.data.startsWith('http://') ||
      item.data.startsWith('https://') ||
      item.data.startsWith('/') ||
      item.data.startsWith('./') ||
      item.data.startsWith('../')
    ) {
      img = `<img src="${props.escapeHtml(
        item.data,
      )}" alt="" style="${props.styleToStringAttribute(cssImg)}" />`;
    }
    //
    else {
      console.log(item.data);
      throw new Error('Bad image data');
    }

    return `<div style="${props.styleToStringAttribute(cssContainer)}">
			${img}
		</div>\n`;
  },

  RenderProperties: function (props) {
    const item = props.item as ImageData;
    const [showModal, setShowModal] = useState<boolean>(false);
    return (
      <>
        <div className={globalStyle.vform}>
          <label htmlFor="img-formula">{Trans('formula')}</label>
          <div className="input-group">
            <span className="input-group-text fst-italic">Æ’</span>
            <InputApplyOnEnter
              id="img-formula"
              value={item.formula}
              onChange={val => props.setItem({ ...item, formula: val })}
            />
          </div>
        </div>

        <div className={globalStyle.vform}>
          <div>
            <label htmlFor="img-url">{Trans('url')}</label>
            <button
              className="btn btn-outline-secondary ms-3"
              onClick={() => setShowModal(true)}
            >
              <FontAwesomeIcon icon={faEllipsisH} />
            </button>
          </div>
          <InputApplyOnEnter
            id="img-url"
            value={item.url}
            onChange={val => props.setItem({ ...item, url: val })}
            placeholder="https://www.example.com/image.jpg"
          />
        </div>

        <div className={globalStyle.section}>{Trans('size')}</div>

        <div className={`${globalStyle.hform} mb-0`}>
          <label htmlFor="width">{Trans('width')}</label>
          <InputApplyOnEnter
            id="width"
            value={item.width || ''}
            onChange={val => {
              const val2: ImageData = { ...item };
              if (val) {
                val2.width = String(val);
              } else {
                delete val2.width;
              }
              return props.setItem(val2);
            }}
            regex={WidthRegex}
          />
        </div>
        <small className="text-muted d-block mb-3">{WidthOptions}</small>

        <div className={`${globalStyle.hform} mb-0`}>
          <label htmlFor="height">{Trans('height')}</label>
          <InputApplyOnEnter
            id="height"
            value={item.height || ''}
            onChange={val => {
              const val2: ImageData = { ...item };
              if (val) {
                val2.height = String(val);
              } else {
                delete val2.height;
              }
              return props.setItem(val2);
            }}
            regex={WidthRegex}
          />
        </div>
        <small className="text-muted d-block mb-3">{WidthOptions}</small>

        <div className={`${globalStyle.hform} mb-0`}>
          <label htmlFor="fit">{Trans('img-fit')}</label>
          <select
            value={
              !!item.fit && !!item.width && !!item.height ? item.fit : 'fill'
            }
            onChange={val => {
              const val2: ImageData = { ...item };
              if (val) {
                val2.fit = val.target.value as any;
              } else {
                delete val2.fit;
              }
              return props.setItem(val2);
            }}
            className="form-select"
            disabled={!item.width || !item.height}
            id="fit"
          >
            <option value="fill">{Trans('img-fit-fill')}</option>
            <option value="contain">{Trans('img-fit-contain')}</option>
            <option value="cover">{Trans('img-fit-cover')}</option>
          </select>
        </div>

        <div className={globalStyle.section}>{Trans('align')}</div>
        <div className="text-center">
          <PropertyAlign
            value={item.align}
            onChange={val => props.setItem({ ...item, align: val })}
          />
        </div>

        <Modal show={showModal} onHide={() => setShowModal(false)} size="lg">
          <Modal.Header closeButton>
            <Modal.Title>Select image</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <FileDialog
              mode="value"
              value={
                item.url && item.url.startsWith('local/')
                  ? item.url.substring(6)
                  : ''
              }
              onChange={val => {
                setShowModal(false);
                props.setItem({
                  ...item,
                  url: val && val.length > 0 ? `local/${val}` : '',
                });
              }}
              api={props.api}
              somethingChanged={() => (imgState += 1)}
            />
          </Modal.Body>
        </Modal>
      </>
    );
  },
};
