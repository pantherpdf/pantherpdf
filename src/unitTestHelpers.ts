/**
 * @file Helper functions used in unit tests
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import { Item, ItemCompiled, Report, ReportCompiled } from './types';
import { defaultWidgets } from './widgets/allWidgets';
import compile from './editor/compile';
import { renderToHtmlContent } from './editor/renderToHtml';

export async function compileComponentTest(
  cmpData: object,
  data: unknown,
): Promise<ItemCompiled> {
  const dt: Report = {
    target: 'pdf',
    name: 'John Johnny',
    children: [cmpData as Item],
    transforms: [],
    properties: {},
    dataUrl: '',
    variables: [],
  };
  const reportCompiled = await compile(dt, data, defaultWidgets, {});
  return reportCompiled.children[0];
}

export async function compileTest(
  report: Report,
  data: unknown,
): Promise<ReportCompiled> {
  return compile(report, data, defaultWidgets, {});
}

export function renderToHtmlContentTest(report: ReportCompiled) {
  return renderToHtmlContent(report, defaultWidgets);
}
