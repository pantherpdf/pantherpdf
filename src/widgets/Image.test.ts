/**
 * @jest-environment node
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023-2024
 * @license MIT
 */

import { Image, ImageData } from './Image';
import { renderWidget } from '../unitTestHelpers';
import { sampleReport } from '../editor/sampleReport';
import type { WidgetNewProps } from './types';

test('Image should render svg', async () => {
  const content = `<svg viewBox="0 0 500 500" xmlns="http://www.w3.org/2000/svg">
  <rect x="37.03" y="43.969" width="420.811" height="419.344" style="fill: rgb(216, 216, 216); stroke: rgb(0, 0, 0);"/>
</svg>`;
  const helper: WidgetNewProps = { report: sampleReport };
  const el = (await Image.newItem(helper)) as ImageData;
  el.url = { formula: `data.generateImg()` };
  const data = {
    generateImg: () => {
      return content;
    },
  };
  const html = await renderWidget(el, data);
  expect(html).toMatchSnapshot();
});
