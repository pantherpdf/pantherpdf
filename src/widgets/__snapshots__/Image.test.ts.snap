// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render svg 1`] = `
<div>
  <div style="display:inline-block;max-width:100%">
    <svg viewbox="0 0 500 500"
         xmlns="http://www.w3.org/2000/svg"
    >
      <rect x="37.03"
            y="43.969"
            width="420.811"
            height="419.344"
            style="fill: rgb(216, 216, 216); stroke: rgb(0, 0, 0);"
      >
      </rect>
    </svg>
  </div>
</div>
`;
